llvm/lib/SYCLLowerIR/LowerESIMD.cpp

```diff
@@ -1272,68 +1286,102 @@ size_t SYCLLowerESIMDPass::runOnFunction(Function &F,
-      continue;
-    }
-    // If vload/vstore is not about the vector-types used by
-    // those globals marked as genx_volatile, We can translate
-    // them directly into generic load/store inst. In this way
-    // those insts can be optimized by llvm ASAP.
-    if (Name.startswith("__esimd_vload")) {
-      if (translateVLoad(*CI, GVTS)) {
+      StringRef Name = Callee->getName();
+
+      // See if the Name represents an ESIMD intrinsic and demangle only if it
+      // does.
+      if (!Name.consume_front(ESIMD_INTRIN_PREF0))
+        continue;
+      // now skip the digits
+      Name = Name.drop_while([](char C) { return std::isdigit(C); });
+
+      // process ESIMD builtins that go through special handling instead of
+      // the translation procedure
+      if (Name.startswith("N2cl4sycl5INTEL3gpu8slm_init")) {
+        // tag the kernel with meta-data SLMSize, and remove this builtin
+        translateSLMInit(*CI);
         ESIMDToErases.push_back(CI);
         continue;
       }
-    }
-    if (Name.startswith("__esimd_vstore")) {
-      if (translateVStore(*CI, GVTS)) {
+      if (Name.startswith("__esimd_pack_mask")) {
+        translatePackMask(*CI);
         ESIMDToErases.push_back(CI);
         continue;
       }
-    }
+      if (Name.startswith("__esimd_unpack_mask")) {
+        translateUnPackMask(*CI);
+        ESIMDToErases.push_back(CI);
+        continue;
+      }
+      // If vload/vstore is not about the vector-types used by
+      // those globals marked as genx_volatile, We can translate
+      // them directly into generic load/store inst. In this way
+      // those insts can be optimized by llvm ASAP.
+      if (Name.startswith("__esimd_vload")) {
+        if (translateVLoad(*CI, GVTS)) {
+          ESIMDToErases.push_back(CI);
+          continue;
+        }
+      }
+      if (Name.startswith("__esimd_vstore")) {
+        if (translateVStore(*CI, GVTS)) {
+          ESIMDToErases.push_back(CI);
+          continue;
+        }
+      }
 
-    if (Name.startswith("__esimd_get_value")) {
-      translateGetValue(*CI);
-      ESIMDToErases.push_back(CI);
-      continue;
-    }
+      if (Name.startswith("__esimd_get_value")) {
+        translateGetValue(*CI);
+        ESIMDToErases.push_back(CI);
+        continue;
+      }
 
-    if (Name.consume_front(SPIRV_INTRIN_PREF)) {
-      translateSpirvIntrinsic(CI, Name, ESIMDToErases);
-      // For now: if no match, just let it go untranslated.
-      continue;
+      if (Name.empty() || !Name.startswith(ESIMD_INTRIN_PREF1))
+        continue;
+      // this is ESIMD intrinsic - record for later translation
+      ESIMDIntrCalls.push_back(CI);
     }
 
-    if (Name.empty() || !Name.startswith(ESIMD_INTRIN_PREF1))
-      continue;
-    // this is ESIMD intrinsic - record for later translation
-    ESIMDIntrCalls.push_back(CI);
+    // Translate loads from SPIRV builtin globals into GenX intrinsics
+    auto *LI = dyn_cast<LoadInst>(&I);
```

Comment:
That's a good point. I agree to reimplement it, but I would prefer to do it in a separate patch if you don't mind. We talked about splitting the LowerESIMD pass into two: ModulePass and FunctionPass. I think it would be appropriate to do those two changes together.

Reply:
OK, sounds good